dali/pipeline/operators/color_space/color_space_conversion.cu

```diff
@@ -0,0 +1,228 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime_api.h>
+#include <utility>
+#include <vector>
+#include "dali/pipeline/operators/color_space/color_space_conversion.h"
+#include "dali/util/npp.h"
+
+namespace dali {
+
+namespace detail {
+
+template<typename T = uint8_t>
+__global__ void ConvertRGBToBGRKernel(const T *input, T *output, unsigned int total_pixels) {
+  unsigned idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= total_pixels) {
+    return;
+  }
+
+  const unsigned int C = 3;
+  const T* pixel_in = &input[idx * C];
+  T* pixel_out = &output[idx * C];
+
+  T tmp[3] = { pixel_in[0], pixel_in[1], pixel_in[2] };
+  pixel_out[0] = tmp[2];
+  pixel_out[1] = tmp[1];
+  pixel_out[2] = tmp[0];
+}
+
+auto ConvertBGRToRGB8uKernel = ConvertRGBToBGRKernel<uint8_t>;
+auto ConvertRGBToBGR8uKernel = ConvertRGBToBGRKernel<uint8_t>;
+
+
+template<typename T = uint8_t>
+__global__ void ConvertGrayToRGBKernel(const T *input, T *output, unsigned int total_pixels) {
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= total_pixels) {
+    return;
+  }
+
+  const T pixel_in = input[idx];
+  const unsigned int C = 3;
+  T* pixel_out = &output[idx * C];
+  pixel_out[0] = pixel_in;
+  pixel_out[1] = pixel_in;
+  pixel_out[2] = pixel_in;
+}
+
+auto ConvertGrayToRGB8uKernel = ConvertGrayToRGBKernel<uint8_t>;
+
+__global__ void ConvertGrayToYCbCr8uKernel(const uint8_t *input, uint8_t *output,
+                                           unsigned int total_pixels) {
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= total_pixels) {
+    return;
+  }
+
+  const uint8_t pixel_in = input[idx];
+  const unsigned int C = 3;
+  uint8_t* pixel_out = &output[idx * C];
+  pixel_out[0] = pixel_in;
+  pixel_out[1] = 128;
+  pixel_out[2] = 128;
+}
+
+template<typename T = uint8_t>
+__global__ void ConvertYCbCrToGrayKernel(const T *input, T *output, unsigned int total_pixels) {
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= total_pixels) {
+    return;
+  }
+
+  const unsigned int C = 3;
+  output[idx] = input[idx * C];
+}
+
+auto ConvertYCbCrToGray8uKernel = ConvertYCbCrToGrayKernel<uint8_t>;
+
+}  // namespace detail
+
+template<>
+void ColorSpaceConversion<GPUBackend>::RunImpl(DeviceWorkspace *ws, const int idx) {
+  const auto &input = ws->Input<GPUBackend>(idx);
+  DALI_ENFORCE(IsType<uint8_t>(input.type()),
+      "Color space conversion accept only uint8 tensors");
+  auto output = ws->Output<GPUBackend>(idx);
+
+  TensorList<CPUBackend> attr_output_cpu;
+
+  int input_C = NumberOfChannels(input_type_);
+  int output_C = NumberOfChannels(output_type_);
+  vector<Dims> input_shape = input.shape();
+  vector<Dims> output_shape = input_shape;
```

Comment:
```suggestion
  const vector<Dims> output_shape = input_shape;
```
same above

Reply:
actually I am editing output_shape so cannot be marked const. I could make input_shape const though